alpha/veneer/semver/semver.go

```diff
@@ -0,0 +1,436 @@
+package semver
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"sort"
+
+	"github.com/blang/semver/v4"
+	"github.com/operator-framework/operator-registry/alpha/action"
+	"github.com/operator-framework/operator-registry/alpha/declcfg"
+	"github.com/operator-framework/operator-registry/alpha/property"
+	"github.com/operator-framework/operator-registry/pkg/image"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/yaml"
+)
+
+// data passed into this module externally
+type Veneer struct {
+	Ref string
+	Reg image.Registry
+}
+
+// IO structs -- BEGIN
+type semverVeneerBundleEntry struct {
+	Image string `json:"image,omitempty"`
+}
+
+type candidateBundles struct {
+	Bundles []semverVeneerBundleEntry `json:"bundles,omitempty"`
+}
+type fastBundles struct {
+	Bundles []semverVeneerBundleEntry `json:"bundles,omitempty"`
+}
+type stableBundles struct {
+	Bundles []semverVeneerBundleEntry `json:"bundles,omitempty"`
+}
+
+type semverVeneer struct {
+	Schema                string           `json:"schema"`
+	GenerateMajorChannels bool             `json:"generateMajorChannels,omitempty"`
+	GenerateMinorChannels bool             `json:"generateMinorChannels,omitempty"`
+	AvoidSkipPatch        bool             `json:"avoidSkipPatch,omitempty"`
+	Candidate             candidateBundles `json:"candidate,omitempty"`
+	Fast                  fastBundles      `json:"fast,omitempty"`
+	Stable                stableBundles    `json:"stable,omitempty"`
+
+	pkg            string `json:"-"` // the derived package name
+	defaultChannel string `json:"-"` // detected "most stable" channel head
+}
+
+// IO structs -- END
+
+// channel "kinds", restricted in this iteration to just these
+const (
+	candidateChannelName string = "candidate"
+	fastChannelName      string = "fast"
+	stableChannelName    string = "stable"
+)
+
+// mapping channel name --> stability, where higher values indicate greater stability
+var channelPriorities = map[string]int{candidateChannelName: 0, fastChannelName: 1, stableChannelName: 2}
+
+// sorting capability for a slice according to the assigned channelPriorities
+type byChannelPriority []string
+
+func (b byChannelPriority) Len() int { return len(b) }
```

Comment:
Do we need a custom `.Len()` reciever on this struct? Why not just use `len()` like normal when we want the length?

Reply:
A certain amount of "monkey see, monkey do" here, as [online](https://thedeveloperblog.com/go/sort-go) [references](https://www.callicoder.com/golang-sorting-custom-functions/) [all](https://stackoverflow.com/questions/30336616/how-to-avoid-re-implementing-sort-interface-for-similar-golang-structs) [say that](https://stackoverflow.com/questions/4947161/is-there-a-way-to-avoid-the-implementation-of-the-full-sort-interface-for-slices). 

Evidently the whole of the `data.Interface` must be implemented for any custom slice sort to be used by `sort.Sort()`.

Reply:
Interesting! Thanks for that, learned something new 😄